BurnpackStore

antimora opened the pull request:
## Pull Request Template

### Checklist

- [ ] Confirmed that `cargo run-checks` command has been executed.
- [ ] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

_Provide links to relevant issues and dependent PRs._

### Changes

_Summarize the problem being addressed and your solution._

### Testing

_Describe how these changes have been tested._

antimora commented:
Added `burn-store/examples/burnpack_inspect.rs` to generate a sample burnpack file to examine. Here is some output:

```
[burn-store]% hexdump -C /tmp/sample.burnpack | head -50
00000000  4e 52 55 42 01 00 d2 01  00 00 a2 67 74 65 6e 73  |NRUB.......gtens|
00000010  6f 72 73 a6 6c 6c 69 6e  65 61 72 31 2e 62 69 61  |ors.llinear1.bia|
00000020  73 a3 65 64 74 79 70 65  63 46 33 32 65 73 68 61  |s.edtypecF32esha|
00000030  70 65 81 18 40 6c 64 61  74 61 5f 6f 66 66 73 65  |pe..@ldata_offse|
00000040  74 73 82 19 80 00 19 81  00 6e 6c 69 6e 65 61 72  |ts.......nlinear|
00000050  31 2e 77 65 69 67 68 74  a3 65 64 74 79 70 65 63  |1.weight.edtypec|
00000060  46 33 32 65 73 68 61 70  65 82 18 80 18 40 6c 64  |F32eshape....@ld|
00000070  61 74 61 5f 6f 66 66 73  65 74 73 82 00 19 80 00  |ata_offsets.....|
00000080  6c 6c 69 6e 65 61 72 32  2e 62 69 61 73 a3 65 64  |llinear2.bias.ed|
00000090  74 79 70 65 63 46 33 32  65 73 68 61 70 65 81 18  |typecF32eshape..|
000000a0  20 6c 64 61 74 61 5f 6f  66 66 73 65 74 73 82 19  | ldata_offsets..|
000000b0  a1 00 19 a1 80 6e 6c 69  6e 65 61 72 32 2e 77 65  |.....nlinear2.we|
000000c0  69 67 68 74 a3 65 64 74  79 70 65 63 46 33 32 65  |ight.edtypecF32e|
000000d0  73 68 61 70 65 82 18 40  18 20 6c 64 61 74 61 5f  |shape..@. ldata_|
000000e0  6f 66 66 73 65 74 73 82  19 81 00 19 a1 00 6c 6c  |offsets.......ll|
000000f0  69 6e 65 61 72 33 2e 62  69 61 73 a3 65 64 74 79  |inear3.bias.edty|
00000100  70 65 63 46 33 32 65 73  68 61 70 65 81 0a 6c 64  |pecF32eshape..ld|
00000110  61 74 61 5f 6f 66 66 73  65 74 73 82 19 a6 80 19  |ata_offsets.....|
00000120  a6 a8 6e 6c 69 6e 65 61  72 33 2e 77 65 69 67 68  |..nlinear3.weigh|
00000130  74 a3 65 64 74 79 70 65  63 46 33 32 65 73 68 61  |t.edtypecF32esha|
00000140  70 65 82 18 20 0a 6c 64  61 74 61 5f 6f 66 66 73  |pe.. .ldata_offs|
00000150  65 74 73 82 19 a1 80 19  a6 80 68 6d 65 74 61 64  |ets.......hmetad|
00000160  61 74 61 a5 66 61 75 74  68 6f 72 6c 42 75 72 6e  |ata.fauthorlBurn|
00000170  20 45 78 61 6d 70 6c 65  6b 64 65 73 63 72 69 70  | Examplekdescrip|
00000180  74 69 6f 6e 78 29 53 61  6d 70 6c 65 20 66 69 6c  |tionx)Sample fil|
00000190  65 20 66 6f 72 20 65 78  61 6d 69 6e 69 6e 67 20  |e for examining |
000001a0  42 75 72 6e 70 61 63 6b  20 66 6f 72 6d 61 74 66  |Burnpack formatf|
000001b0  66 6f 72 6d 61 74 68 62  75 72 6e 70 61 63 6b 68  |formathburnpackh|
000001c0  70 72 6f 64 75 63 65 72  64 62 75 72 6e 67 76 65  |producerdburngve|
000001d0  72 73 69 6f 6e 66 30 2e  31 39 2e 30 5a 7e 3e 3d  |rsionf0.19.0Z~>=|
000001e0  e6 00 93 bd 14 7a 89 bc  d2 91 57 3d fa c4 4a 3d  |.....z....W=..J=|
000001f0  20 05 6d 3b d6 de 7e 3d  42 5c 73 3d 70 9b da bc  | .m;..~=B\s=p...|
00000200  61 e6 83 3d c0 d3 d7 3b  62 5a 44 3d 2c 44 b9 3c  |a..=...;bZD=,D.<|
00000210  a5 8c 1c bd 11 83 a7 3d  93 a1 0b bd bd 06 a2 3d  |.......=.......=|
00000220  00 30 af b9 47 7e b4 bd  64 e9 02 bd 20 eb fb bb  |.0..G~..d... ...|
00000230  48 38 b9 3c 35 0c 8d 3d  3a c1 1b 3d 30 26 ee bb  |H8.<5..=:..=0&..|
00000240  b0 be e4 3c d8 fd d4 3c  3c 48 18 bd 40 99 f4 3a  |...<...<<H..@..:|
00000250  9b b1 4c bd 5e 39 27 3d  f1 fa a3 3d e8 db c7 3c  |..L.^9'=...=...<|
00000260  88 91 ef 3c 6b f4 7c bd  a2 38 a9 bd ab cf 89 bd  |...<k.|..8......|
00000270  f5 1f 8a bd 33 3e a9 3d  30 95 29 3c 4b 99 8a 3d  |....3>.=0.)<K..=|
00000280  5e 4b a5 bd 68 d2 20 bc  d0 ed 00 3c 8a e7 28 3d  |^K..h. ....<..(=|
00000290  b0 30 0b 3c c2 b8 17 3d  6d 41 8a 3d 69 ad ae bd  |.0.<...=mA.=i...|
000002a0  e8 fe 0c 3d b1 50 b4 bd  57 2a ac 3d ce d8 1a 3d  |...=.P..W*.=...=|
000002b0  3a a3 64 3d b4 be 34 bd  80 2f a4 3b 6c 61 92 bd  |:.d=..4../.;la..|
000002c0  38 e0 e3 bc a0 f4 89 bb  73 4e 4d bd 43 8b 9f bd  |8.......sNM.C...|
000002d0  e0 7e 14 3d d1 e7 4b bd  50 9b 3e 3c c1 7b b3 bd  |.~.=..K.P.><.{..|
000002e0  48 d8 b5 bc c7 c3 a2 3d  c0 5f 53 bb eb 0d a5 bd  |H......=._S.....|
000002f0  30 a3 f8 bc f7 04 a5 3d  90 dd a9 bd fb e0 9a 3d  |0......=.......=|
00000300  a0 1d 83 3b 24 f3 aa bc  b0 09 d2 bc 54 34 aa bd  |...;$.......T4..|
00000310  4e 0e 27 bd 00 91 a8 3c  75 be b1 3d 6c 99 85 3c  |N.'....<u..=l..<|
[burn-store]%
```

antimora commented:
Rebased with the main; @laggui @nathanielsimard  Ready for review.

laggui reviewed (commented):
(this is a bit more of a high level review)

There seems to be a couple of dead code paths that are just marked as `#[allow(dead_code)]`. Not sure why there is "legacy" code in new additions 😄 

Also, the memory usage of the CPU backend seems to point to some additional allocations overhead that we should try to minimze. The detailed benchmarks seem to point in that direction as well.

The current recorder approach (read all at once) is really optimized for speed, but the gap in read/write speeds should still not be as wide as we see in the benchmarks. I think we can optimize the sequential lazy read path (the common use case when loading a model) to improve this. My guess is that there are too many syscalls right now due to the seek + read for each tensor. We should probably separate the IO strategy from the tensor/snapshot object. I am not sure how it would be best yet but just giving my thoughts.

antimora commented:
@laggui I will investigate issue. 

Looking at the benchmark results, BurnpackStore definitely calls more alloc than NamedMpk, even though it uses less memory overall.

<img width="1528" height="770" alt="image" src="https://github.com/user-attachments/assets/bca817be-4f9b-420e-bc12-4fdf68ab540b" />

NamedMpk reads a file and loads all its content into memory. Then it transfers vec ownership to Data without copying again. 

BurnpackStore uses `TensorData::from_bytes_vec()`; I need to see if there is more efficient API.

antimora commented:
@laggui UPDATE:

 The 10x slowdown is caused by:

  1. LargeModel::new(&device) creates Params with random weights (but lazy via OnceCell)
  2. apply_from() → module.map() → Param::map() → .consume() → .val() forces 360MB of random weights to materialize!
  3. Then TensorSnapshots create another 327MB of loaded weights
  4. Total waste: 360MB + overhead

antimora commented:
@laggui 
> Also, the memory usage of the CPU backend seems to point to some additional allocations overhead that we should try to minimze. The detailed benchmarks seem to point in that direction as well.

I have figured out the root cause and additional issues:

1. ModuleMapper was initializing params instead of replacing uninitialized params. I have fixed this. Additionally partially loaded model keeps skipped params intact. 
2. Params also had on_save/on_load hooks that Nathaniel added recently for record saving which we totally forgot. Now Store's applier/collector properly execute those hooks - take the same execution path as the recorder. 

I updated the benchmark table in the PR description. BurnpackStore performs the best of all now. PytorchStore and SafetensorsStore also benefit from these fixes.

nathanielsimard reviewed (commented):
The file design seems good, and the overall architecture seems flexible. I have quite a few comments, mostly related to naming and documentation.

I think a lot of comments are quite bad; they describe how things work (sometimes not even accurately) instead of documenting how the code should be used.

Also, some formatting isn’t correct. We shouldn’t use `**title**` but `# title in` most places (`# Returns`, `# Safety`, etc.).
